src/DataStructures/CachedTempBuffer.hpp

```diff
@@ -0,0 +1,59 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <utility>  // IWYU pragma: keep  // for std::move
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/TempBuffer.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/*!
+ * \ingroup DataStructuresGroup
+ * A temporary buffer with contents computed on demand.
+ *
+ * When an entry in the buffer is first requested, it is computed by
+ * the `Computer` class.  Subsequent requests just return the cached
+ * value.  The computer can itself request data from the cache to use
+ * in its computations.
+ *
+ * For the cache
+ * \snippet Test_CachedTempBuffer.cpp alias
+ * the function used to compute `Tags::Scalar2<DataType>` is
+ * \snippet Test_CachedTempBuffer.cpp compute_func
+ */
+template <typename Computer, typename... Tags>
+class CachedTempBuffer {
+ public:
+  /// Construct the buffer with the given computer.  `size` is passed
+  /// to the underlying `TempBuffer` constructor.
+  CachedTempBuffer(const size_t size, Computer computer) noexcept
+      : data_(size), computer_(std::move(computer)) {}
+
+  /// Obtain a value from the buffer, computing it if necessary.
+  template <typename Tag>
+  const db::const_item_type<Tag>& get_var(Tag /*meta*/) noexcept {
+    // This function can't be called "get" because that interferes
+    // with the ADL needed to access data_.
+    if (not get<Computed<Tag>>(computed_flags_)) {
+      computer_(make_not_null(&get<Tag>(data_)), make_not_null(this), Tag{});
```

Comment:
note from call: `using tuples::get` could work

Reply:
It sort of works.  Importing `::get` and `tuples::get` will let this function be renamed to `get`, but then some other resolution rule forces all the uses in IntermediateVars to change from `get_var()` to `this->get()` (and `::get` has to be imported in a bunch of places).  It is not clear to me whether or not this is an improvement, so I'm going to leave it as it.

Reply:
ok thanks for trying